Enable logf128 constant folding for hosts with 128bit long double

MDevereau opened the pull request:
Hosts with a long double size of 128 bits can benefit from constant fp128 folding. GCC versions more recent than version 12 must use the _Float128 type for logf128.

MDevereau committed:
Enable logf128 constant folding for AArch64

AArch64 has a long double bit length of 128. Therefore, it can benefit
from constant fp128 folding. GCC versions more recent than version 12
must use the _Float128 type for logf128.

efriedma-quic commented:
I don't understand how this works.  logf128 is defined to take a _Float128, by the C standard.  long double is not _Float128, and is not compatible with _Float128.  Therefore, it should not be possible to implicitly convert a `_Float128(*)(_Float128)` to `long double (*)(long double)`.

davemgreen commented on llvm/include/llvm/Support/float128.h:
```diff
@@ -11,7 +11,14 @@
 namespace llvm {
 
-#if defined(__clang__) && defined(__FLOAT128__) &&                             \
+#if defined(__aarch64__)
```
I would expect we might need to be careful of systems that have long-double==double, such as windows and darwin targets.

MDevereau replied:
I've added a check for the mantissa size now which should help with this.

davemgreen reviewed (commented):
> logf128 is defined to take a _Float128, by the C standard. long double is not _Float128, and is not compatible with _Float128. Therefore, it should not be possible to implicitly convert a _Float128(*)(_Float128) to long double (*)(long double).

Hi - Can you explain more about how they are not compatible? Do you just mean at a type/C level? I thought that if they were the same size they would operate in the same way (i.e logl is just an alias to logf128 in libm on the system I have).

MDevereau commented:
>I don't understand how this works. logf128 is defined to take a _Float128, by the C standard. long double is not _Float128, and is not compatible with _Float128. Therefore, it should not be possible to implicitly convert a _Float128(*)(_Float128) to long double (*)(long double).

This was the rationale for the typedef of _Float128 to float128 for gcc versions higher than 12 and why the changes in ConstantFolding.cpp were warranted. It appears pre gcc13, logf128 did not use the _Float128 type but rather an alias of long double or __float128. What this typedef does is remove the implicit conversion which gives an error message.
https://godbolt.org/z/5bdneMz9o
https://godbolt.org/z/fMf8oKor1

You can see this was not an issue in gcc12
https://godbolt.org/z/vs5eaxvE9

efriedma-quic commented:
Oh, I see, gcc 13 has _Float128, and on older versions glibc typedefs it to long double.

Can we make this not AArch64-specific?  It seems like it affects any target where long double is 128 its.

MDevereau commented:
> Oh, I see, gcc 13 has _Float128, and on older versions glibc typedefs it to long double.
> 
> Can we make this not AArch64-specific? It seems like it affects any target where long double is 128 its.

I'm not opposed to the idea of making this non AArch64-specific, but on further inspection it seems logf128 using _Float128/__float128/long double isn't consistent with GCC/Clang versions across hosts.

While trying to figure out which host/compiler combinations were appropriate to use, it quickly got quite confusing. Including math.h seems to typedef long double to _Float128 under certain circumstances, but support for the original untypedef'd types seems to vary.
I ran this test (https://godbolt.org/z/f1c9q9jh4) while substituting the type to check which compiler/host combination works for the types without including math.h. I'm interested in AArch64/x86 supporting this predominantly so I've included those:

AArch64:
| Type | GCC-13.1 | GCC-12.4 | Clang-18.1 |
| ---------- |  ---------- | ---------- | ---------- |
| _Float128 | pass | fail | fail |
| __float128 | fail | fail | fail |
| long double | pass | pass | pass |

x86:
| Type | GCC-13.1 | GCC-12.4 | Clang-18.1 |
| ---------- |  ---------- | ---------- | ---------- |
| _Float128 | pass | fail | fail |
| __float128 | pass | pass | pass |
| long double | pass | pass | pass |
 
Where it gets interesting is when using these types as parameters to logf128 from math.h. In GCC-13 and over logf128 appears to be a builtin which takes _Float128 and does not allow implicit conversions, but on x86 logf128 seems to take a typedef of _Float128/__float128 which throws when given long double. I then ran this logf128 function pointer test (https://godbolt.org/z/GvM78hbbh) for the same host/target combinations and got these results:

AArch64:
| Type | GCC-13.1 | GCC-12.4 | Clang-18.1 |
| ---------- |  ---------- | ---------- | ---------- |
| _Float128 | pass | pass | pass |
| __float128 | fail | fail | fail |
| long double | fail | pass | pass |

x86:
| Type | GCC-13.1 | GCC-12.4 | Clang-18.1 |
| ---------- |  ---------- | ---------- | ---------- |
| _Float128 | pass | pass | fail |
| __float128 | fail | pass | fail |
| long double | fail | fail | fail |

So while x86 has a long double length of 128 bits, I don't think it's ok to enable this based on long double information alone. It looks like the only way to stay sane in this scenario is to do it on a per host and per compiler basis.

efriedma-quic commented:
I'd still prefer to use some sort of detection. If you can't figure out a way to do it in the header itself, please use CMake tests.

MDevereau committed:
Remove aarch64 constraint

Also enable optimisation without cmake opt-in

MDevereau committed:
Enable optimization without opt-in

MDevereau commented:
> I'd still prefer to use some sort of detection. If you can't figure out a way to do it in the header itself, please use CMake tests.

I've added detection of long double size. Confusingly some x86 targets defined long double as 16 bytes but are actually fp80 types, so I've checked the size of the mantissa too. I've also removed the CMake opt-in and detect logf128 by default now.

efriedma-quic commented on llvm/include/llvm/Support/float128.h:
```diff
@@ -11,7 +11,15 @@
+#if !defined(__LONG_DOUBLE_IBM128__) && (__SIZEOF_LONG_DOUBLE__ == 16) &&      \
+    (__SIZEOF_INT128__ == 16) && (__LDBL_MANT_DIG__ == 113)
+#define HAS_IEE754_FLOAT128
+#if (defined(__GNUC__) && __GNUC__ > 12)
```
Instead of explicitly checking the gcc version, can we `typedef decltype(logf128(0.)) float128;` or something like that?

MDevereau replied:
This would require including math.h in this file (which isn't a problem but worth mentioning). It's also possible for hosts to satisfy the condition and fall through to this but not have logf128 ([I've seen x86 with clang do this](https://godbolt.org/z/hcn87xaon)). It's possible to add HAS_LOGF128 to the check, which is set at CMake time in order to get around this, but this isn't really about logf128 and more about the 128 bit floating-point types themselves. Granted, my logf128 patches are the only thing using this, but it still seems incorrect. 

Apart from this, it does function well though.

efriedma-quic replied:
I'm mostly concerned that checking for __GNUC__ is overly specific, and might break on compilers you haven't seen, or compilers which don't exist yet (e.g. if clang's support for _Float128 changes).

MDevereau replied:
I've instead decided to move to a CMake test which checks the availability of logf128, the availability of a true 128bit float type and that type's compatibility with the logf128 symbol to receive accurate answers instead of trying to do this inside a header file. 

A problem I faced when trying to run this on an x86 machine was that the size of long doubles were defined to be 16 bytes (when they are actually fp80) however calling logf128 with this type caused error codes to be returned from the function while compiling without issue. Using __float128 which is guaranteed to perform a 128 bit calculation even though the machine uses fp80 works OK though. I'm seeing both x86 with fp80 and aarch64 compile and run this without issues now. Instead of explicitly checking the GCC version, the _Float128 type which is available in GCC12+ is now asserted for availability and correctness and used if it is available.

MDevereau committed:
Check for viable logf128 and use that for float128's typedef

MDevereau commented on llvm/cmake/config-ix.cmake:
```diff
@@ -271,6 +260,64 @@ if(C_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW)
+file(WRITE "${CMAKE_CURRENT_BINARY_DIR}/logf128_long_double.cpp"
+"
+extern \"C\" {
+long double logf128(long double);
```
The reason I ended up doing things this way is because Clang on x86 doesn't pick up the logf128 symbol when including cmath. Clang just doesn't seem to have the magic definitions to unlock it like GCC does. It's probably okay to include cmath for long double and _Float128, but for clang I think we do need to explicitly declare the function prototype to get it working on x86.

efriedma-quic replied:
For the initial patch, let's just handle the case where logf128 is actually declared in math.h.  The case where glibc actually provides logf128, but doesn't declare it, is a weird edge case, and I'd prefer to handle it in a followup patch.

MDevereau commented on llvm/cmake/config-ix.cmake:
```diff
@@ -271,6 +260,64 @@ if(C_SUPPORTS_WERROR_UNGUARDED_AVAILABILITY_NEW)
+  unset(LOGF128_TEST_RUN CACHE)
+  unset(LOGF128_TEST_COMPILE CACHE)
+  try_run(
+    LOGF128_TEST_RUN
```
x86 hosts with 80bit long doubles can compile logf128 but will return an error code. `__LDBL_MANT_DIG__` does work fine for hosts with 128bit long doubles, but this will also disable the fold for x86 targets with fp80 that can still perform an accurate fold with __float128 types. I'm struggling to think of an alternative way to select the most valid and best type at compile time to use without try_run based on the following criteria:

Logf128 symbol availability
Logf128 function prototype availability
Long double total size (Needs to be 16 bytes)
Long double mantissa size (Rules out fp80)
GCC version(12+ won't accept anything other than _Float128 for logf128)
If we're compiling with GCC or Clang (Clang needs to declare the logf128 prototype)
__float128 availability
_Float128 availability

efriedma-quic replied:
You only need to check `__LDBL_MANT_DIG__` if you're trying to use the type `long double`.  You can assume `__float128` and `_Float128` have the right format.

MDevereau replied:
Firstly, thanks for your patience with this review. 

I'm still unsure how to differentiate in a header file between `_Float128` and `long double` for GCC12+, where trying to pass in anything other than GCC 12+'s dedicated _Float128 type will cause a compilation error. A condition along the lines of
```c++
#if (__LDBL_MANT_DIG__ == 113) && !defined(__LONG_DOUBLE_IBM128__) && (__SIZEOF_INT128__ == 16)
typedef long double float128;
#else
```
will be valid for GCC12 and GCC11 while it is actually only valid for GCC11: long double isn't valid for logf128 anymore after GCC12, but the host still has a long double mantissa size of 113.

MDevereau committed:
Remove cmake tests

MDevereau committed:
Lit uses LLVM_HAS_LOGF128 to run tests and is no longer a command line option

efriedma-quic commented on llvm/include/llvm/Support/float128.h:
```diff
@@ -9,18 +9,17 @@
-    !defined(__LONG_DOUBLE_IBM128__) &&                                        \
-    (defined(__GNUC__) || defined(__GNUG__))
+#ifdef HAS_LOGF128
+#if (__LDBL_MANT_DIG__ == 113) && !defined(__LONG_DOUBLE_IBM128__) &&          \
```
Is the check for __LDBL_MANT_DIG__ necessary with the current version?

Otherwise, this approach seems fine.

efriedma-quic replied:
(I was suggesting __LDBL_MANT_DIG__ with the old version that explicitly wrote out the type "long double".  It should be safe to assume that if logf128 returns a long double, then long double is actually an IEEE 128-bit float.)

MDevereau replied:
Sorry for the late update, I've been away for a few weeks. I believe its unnecessary in this version, if `HAS_LOGF128` is true the logf128 function prototype should return a valid IEEE 128-bit float.

I believe this will expose a problem I've seen where x86 targets with fp80 will return a slightly different result from AArch64. I.e., AArch64's `0xL00000000000000007FFF800000000000` vs x86's `0xL0000000000000000FFFF800000000000` for the Constant folding test log_e_negative_infinity and a few others. I'm not sure if this is due to a target specific implementation of logf128 currently.

MDevereau committed:
Remove check for __LDBL_MANT_DIG__

efriedma-quic reviewed (approved):
LGTM

MDevereau committed:
Merge branch 'main' into aarch64-logf128

nikic commented:
Reverted to fix the buildbot failures.

Before reapplying this, please also make sure that you fix the 1.5% regression to clang build time this causes. Presumably this is due to including `<cmath>` in a core header. Probably most easily tested using the DivisionByConstantInfo.cpp file.

kraj commented:
@MDevereau sadly the relanded patch has caused regression while compiling gawk on aarch64, I am attaching the failing test file and script to run

```
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /home/khem/llvm-project/build/bin/clang -DGAWK -DHAVE_CONFIG_H -I./.. -I. -I.. -g -O2 -DNDEBUG -DNDEBUG -g -O2 -DNDEBUG -DNDEBUG -MT pma.o -MD -MP -MF .deps/pma.Tpo -c -o pma.o pma.c
1.      <eof> parser at end of file
2.      Optimizer
3.      Running pass "require<globals-aa>,function(invalidate<aa>),require<profile-summary>,cgscc(devirt<4>(inline,function-attrs<skip-non-recursive-function-attrs>,openmp-opt-cgscc,function<eager-inv;no-rerun>(sroa<modify-cfg>,early-cse<memssa>,speculative-execution<only-if-divergent-target>,jump-threading,correlated-propagation,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,aggressive-instcombine,libcalls-shrinkwrap,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,reassociate,constraint-elimination,loop-mssa(loop-instsimplify,loop-simplifycfg,licm<no-allowspeculation>,loop-rotate<header-duplication;no-prepare-for-lto>,licm<allowspeculation>,simple-loop-unswitch<no-nontrivial;trivial>),simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,loop(loop-idiom,indvars,simple-loop-unswitch<no-nontrivial;trivial>,loop-idiom-vectorize,loop-deletion,loop-unroll-full),sroa<modify-cfg>,vector-combine,mldst-motion<no-split-footer-bb>,gvn<>,sccp,bdce,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,jump-threading,correlated-propagation,adce,memcpyopt,dse,move-auto-init,loop-mssa(licm<allowspeculation>),coro-elide,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;hoist-common-insts;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>),function-attrs,function(require<should-not-run-function-passes>),coro-split)),function(invalidate<should-not-run-function-passes>),cgscc(devirt<4>())" on module "pma.c"
4.      Running pass "cgscc(devirt<4>(inline,function-attrs<skip-non-recursive-function-attrs>,openmp-opt-cgscc,function<eager-inv;no-rerun>(sroa<modify-cfg>,early-cse<memssa>,speculative-execution<only-if-divergent-target>,jump-threading,correlated-propagation,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,aggressive-instcombine,libcalls-shrinkwrap,tailcallelim,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,reassociate,constraint-elimination,loop-mssa(loop-instsimplify,loop-simplifycfg,licm<no-allowspeculation>,loop-rotate<header-duplication;no-prepare-for-lto>,licm<allowspeculation>,simple-loop-unswitch<no-nontrivial;trivial>),simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,loop(loop-idiom,indvars,simple-loop-unswitch<no-nontrivial;trivial>,loop-idiom-vectorize,loop-deletion,loop-unroll-full),sroa<modify-cfg>,vector-combine,mldst-motion<no-split-footer-bb>,gvn<>,sccp,bdce,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>,jump-threading,correlated-propagation,adce,memcpyopt,dse,move-auto-init,loop-mssa(licm<allowspeculation>),coro-elide,simplifycfg<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;hoist-common-insts;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,instcombine<max-iterations=1;no-use-loop-info;no-verify-fixpoint>),function-attrs,function(require<should-not-run-function-passes>),coro-split))" on module "pma.c"
 #0 0x0000c71975749f44 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/khem/llvm-project/build/bin/clang+0x35d9f44)
 #1 0x0000c71975748450 llvm::sys::CleanupOnSignal(unsigned long) (/home/khem/llvm-project/build/bin/clang+0x35d8450)
 #2 0x0000c7197569cf18 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x0000e517c22828f8 (linux-vdso.so.1+0x8f8)
 #4 0x0000c7197487d08c (anonymous namespace)::ConstantFoldScalarCall1(llvm::StringRef, unsigned int, llvm::Type*, llvm::ArrayRef<llvm::Constant*>, llvm::TargetLibraryInfo const*, llvm::CallBase const*) (.isra.0) ConstantFolding.cpp:0:0
 #5 0x0000c719748d19ec (anonymous namespace)::CallAnalyzer::visitCallBase(llvm::CallBase&) InlineCost.cpp:0:0
 #6 0x0000c719748d40f4 (anonymous namespace)::CallAnalyzer::analyze() (.part.0) InlineCost.cpp:0:0
 #7 0x0000c719748d5acc llvm::getInlineCost(llvm::CallBase&, llvm::Function*, llvm::InlineParams const&, llvm::TargetTransformInfo&, llvm::function_ref<llvm::AssumptionCache& (llvm::Function&)>, llvm::function_ref<llvm::TargetLibraryInfo const& (llvm::Function&)>, llvm::function_ref<llvm::BlockFrequencyInfo& (llvm::Function&)>, llvm::ProfileSummaryInfo*, llvm::OptimizationRemarkEmitter*) (/home/khem/llvm-project/build/bin/clang+0x2765acc)
 #8 0x0000c719748d5e70 llvm::getInlineCost(llvm::CallBase&, llvm::InlineParams const&, llvm::TargetTransformInfo&, llvm::function_ref<llvm::AssumptionCache& (llvm::Function&)>, llvm::function_ref<llvm::TargetLibraryInfo const& (llvm::Function&)>, llvm::function_ref<llvm::BlockFrequencyInfo& (llvm::Function&)>, llvm::ProfileSummaryInfo*, llvm::OptimizationRemarkEmitter*) (/home/khem/llvm-project/build/bin/clang+0x2765e70)
 #9 0x0000c71978ece8e4 llvm::InlineCost llvm::function_ref<llvm::InlineCost (llvm::CallBase&)>::callback_fn<getDefaultInlineAdvice(llvm::CallBase&, llvm::AnalysisManager<llvm::Function>&, llvm::InlineParams const&)::'lambda2'(llvm::CallBase&)>(long, llvm::CallBase&) InlineAdvisor.cpp:0:0
#10 0x0000c71978ed38e0 llvm::shouldInline(llvm::CallBase&, llvm::function_ref<llvm::InlineCost (llvm::CallBase&)>, llvm::OptimizationRemarkEmitter&, bool) (/home/khem/llvm-project/build/bin/clang+0x6d638e0)
#11 0x0000c71978ed48f0 getDefaultInlineAdvice(llvm::CallBase&, llvm::AnalysisManager<llvm::Function>&, llvm::InlineParams const&) InlineAdvisor.cpp:0:0
#12 0x0000c71978ed4998 llvm::DefaultInlineAdvisor::getAdviceImpl(llvm::CallBase&) (/home/khem/llvm-project/build/bin/clang+0x6d64998)
#13 0x0000c71978ed1650 llvm::InlineAdvisor::getAdvice(llvm::CallBase&, bool) (/home/khem/llvm-project/build/bin/clang+0x6d61650)
#14 0x0000c71976a09bcc llvm::InlinerPass::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/khem/llvm-project/build/bin/clang+0x4899bcc)
#15 0x0000c71976861fdc llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::InlinerPass, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/khem/llvm-project/build/bin/clang+0x46f1fdc)
#16 0x0000c71974861424 llvm::PassManager<llvm::LazyCallGraph::SCC, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/khem/llvm-project/build/bin/clang+0x26f1424)
#17 0x0000c7197685dd5c llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::PassManager<llvm::LazyCallGraph::SCC, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/khem/llvm-project/build/bin/clang+0x46edd5c)
#18 0x0000c71974867580 llvm::DevirtSCCRepeatedPass::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/khem/llvm-project/build/bin/clang+0x26f7580)
#19 0x0000c7197685ddbc llvm::detail::PassModel<llvm::LazyCallGraph::SCC, llvm::DevirtSCCRepeatedPass, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&>::run(llvm::LazyCallGraph::SCC&, llvm::AnalysisManager<llvm::LazyCallGraph::SCC, llvm::LazyCallGraph&>&, llvm::LazyCallGraph&, llvm::CGSCCUpdateResult&) (/home/khem/llvm-project/build/bin/clang+0x46eddbc)
#20 0x0000c7197486323c llvm::ModuleToPostOrderCGSCCPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/khem/llvm-project/build/bin/clang+0x26f323c)
#21 0x0000c7197685dcfc llvm::detail::PassModel<llvm::Module, llvm::ModuleToPostOrderCGSCCPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/khem/llvm-project/build/bin/clang+0x46edcfc)
#22 0x0000c719751a4b7c llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/khem/llvm-project/build/bin/clang+0x3034b7c)
#23 0x0000c71976a084e8 llvm::ModuleInlinerWrapperPass::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/khem/llvm-project/build/bin/clang+0x48984e8)
#24 0x0000c7197685d24c llvm::detail::PassModel<llvm::Module, llvm::ModuleInlinerWrapperPass, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/khem/llvm-project/build/bin/clang+0x46ed24c)
#25 0x0000c719751a4b7c llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/home/khem/llvm-project/build/bin/clang+0x3034b7c)
#26 0x0000c719759911c0 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&, clang::BackendConsumer*) BackendUtil.cpp:0:0
#27 0x0000c719759939e0 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>, clang::BackendConsumer*) (/home/khem/llvm-project/build/bin/clang+0x38239e0)
#28 0x0000c71975ee4484 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/khem/llvm-project/build/bin/clang+0x3d74484)
#29 0x0000c7197764076c clang::ParseAST(clang::Sema&, bool, bool) (/home/khem/llvm-project/build/bin/clang+0x54d076c)
#30 0x0000c71976136ca0 clang::FrontendAction::Execute() (/home/khem/llvm-project/build/bin/clang+0x3fc6ca0)
#31 0x0000c719760c40f0 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/khem/llvm-project/build/bin/clang+0x3f540f0)
#32 0x0000c719761eed10 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/khem/llvm-project/build/bin/clang+0x407ed10)
#33 0x0000c7197301b9c0 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/khem/llvm-project/build/bin/clang+0xeab9c0)
#34 0x0000c71973015afc ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#35 0x0000c71975f26260 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#36 0x0000c7197569d3c4 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/khem/llvm-project/build/bin/clang+0x352d3c4)
#37 0x0000c71975f265ec clang::driver::CC1Command::Execute(llvm::ArrayRef<std::optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#38 0x0000c71975ef340c clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/khem/llvm-project/build/bin/clang+0x3d8340c)
#39 0x0000c71975ef3c34 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/khem/llvm-project/build/bin/clang+0x3d83c34)
#40 0x0000c71975f03724 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/khem/llvm-project/build/bin/clang+0x3d93724)
#41 0x0000c719730189d4 clang_main(int, char**, llvm::ToolContext const&) (/home/khem/llvm-project/build/bin/clang+0xea89d4)
#42 0x0000c71972f43adc main (/home/khem/llvm-project/build/bin/clang+0xdd3adc)
#43 0x0000e517c1c773fc (/lib/aarch64-linux-gnu/libc.so.6+0x273fc)
#44 0x0000e517c1c774cc __libc_start_main (/lib/aarch64-linux-gnu/libc.so.6+0x274cc)
#45 0x0000c719730155b0 _start (/home/khem/llvm-project/build/bin/clang+0xea55b0)
clang: error: clang frontend command failed with exit code 139 (use -v to see invocation)
clang version 20.0.0git (git@github.com:kraj/llvm-project f67388232384682fb442d6e5501d9259c41fd714)
Target: aarch64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/khem/llvm-project/build/bin
```

[pma.zip](https://github.com/user-attachments/files/16734930/pma.zip)

davemgreen commented:
Thanks for the report. It looks like there is an unguarded use of TLI that might have been nullptr from certain passes. I've added 83a5c7cb62e404a713a35445b755cf0109650279 that hopefully addresses it.

kraj commented:
> Thanks for the report. It looks like there is an unguarded use of TLI that might have been nullptr from certain passes. I've added [83a5c7c](https://github.com/llvm/llvm-project/commit/83a5c7cb62e404a713a35445b755cf0109650279) that hopefully addresses it.

thats for quick turn around. this patch fixed the problem
